packages/web-api-scan-runner/src/crawl-runner/scan-feed-generator.ts

```diff
@@ -0,0 +1,101 @@
+        @inject(GuidGenerator) private readonly guidGenerator: GuidGenerator,
+        @inject(GlobalLogger) private readonly logger: GlobalLogger,
+    ) {}
+
+    public async queueDiscoveredPages(websiteScanResult: WebsiteScanResult, pageScanResult: OnDemandPageScanResult): Promise<void> {
+        return this.retryHelper.executeWithRetries(
+            () => this.queueDiscoveredPagesImpl(websiteScanResult, pageScanResult),
+            async (error: Error) => {
+                this.logger.logError(`Failure to queue discovered pages to scan. Retrying on error.`, {
+                    error: System.serializeError(error),
+                });
+            },
+            this.maxRetryCount,
+            1000,
+        );
+    }
+
+    private async queueDiscoveredPagesImpl(websiteScanResult: WebsiteScanResult, pageScanResult: OnDemandPageScanResult): Promise<void> {
+        const urlsToScan = this.getUrlsToScan(websiteScanResult);
+        if (urlsToScan.length > 0) {
+            this.logger.logInfo(`Discovered ${urlsToScan.length} new pages to scan.`, {
+                discoveredUrls: JSON.stringify(urlsToScan),
+            });
+        } else {
+            this.logger.logInfo(`Discovered no new pages to scan.`);
+
+            return;
+        }
+
+        const batchId = this.guidGenerator.createGuid();
+        const scanRequests = this.createScanRequests(batchId, urlsToScan, pageScanResult, websiteScanResult);
+        const pageScans = scanRequests.map((scanRequest) => {
+            return { scanId: scanRequest.scanId, url: scanRequest.url, timestamp: new Date().toJSON() };
+        });
+        const updatedWebsiteScanResult: Partial<WebsiteScanResult> = {
+            id: websiteScanResult.id,
+            pageScans,
+            _etag: websiteScanResult._etag,
+        };
+        await this.websiteScanResultProvider.mergeOrCreate(updatedWebsiteScanResult);
+        await this.scanDataProvider.writeScanRunBatchRequest(batchId, scanRequests);
+        this.logger.logInfo(`Discovered pages has been queued for scanning.`);
+    }
+
+    private createScanRequests(
+        batchId: string,
+        urls: string[],
+        pageScanResult: OnDemandPageScanResult,
+        websiteScanResult: WebsiteScanResult,
+    ): ScanRunBatchRequest[] {
+        return urls.map((url) => {
+            // preserve GUID origin for a single batch scope
+            const scanId = this.guidGenerator.createGuidFromBaseGuid(batchId);
+            this.logger.logInfo('Generated new scan id for the discovered page.', {
+                batchId,
+                discoveredScanId: scanId,
+                discoveredUrl: url,
+            });
+
+            return {
+                scanId,
+                url,
+                priority: _.isNil(pageScanResult.priority) ? 0 : pageScanResult.priority,
+                deepScan: true,
+                ...(_.isEmpty(pageScanResult.scanNotifyUrl) ? {} : { scanNotifyUrl: pageScanResult.scanNotifyUrl }),
```

Comment:
nit: This line is a bit confusing to look at, could it be extracted into a variable?

Reply:
updated